metsim/methods/mtclim.py

```diff
@@ -19,197 +19,125 @@
-    dtr = df['t_max'] - df['t_min']
-    df['dtr'] = dtr
-    sm_dtr = df['smoothed_dtr']
-    df['tfmax'] = _calc_tfmax(df['prec'], dtr, sm_dtr)
-    tdew = df.get('tdew', df['t_min'])
-    pva = df.get('hum', svp(tdew.values))
-    pa = atm_pres(elev, params['lapse_rate'])
-    yday = df.index.dayofyear - 1
-    df['dayl'] = sg['daylength'][yday]
-
-    # Calculation of tdew and shortwave. tdew is iterated on until
-    # it converges sufficiently
-    tdew_old = tdew
-    tdew, pva = sw_hum_iter(df, sg, pa, pva, dtr, params)
-    while(np.sqrt(np.mean((tdew-tdew_old)**2)) > params['tdew_tol']):
-        tdew_old = np.copy(tdew)
-        tdew, pva = sw_hum_iter(df, sg, pa, pva, dtr, params)
-    df['vapor_pressure'] = pva
-
-
-def sw_hum_iter(df: pd.DataFrame, sg: dict, pa: float, pva: pd.Series,
-                dtr: pd.Series, params: dict):
-    """
-    Calculated updated values for dewpoint temperature
-    and saturation vapor pressure.
+def _calc_tdew_iter(tdew_temp, yday, tfmax, t_day, dayl, t_min,
+                    seasonal_prec, dtr, params, sg):
+    '''helper function to facilitate calculating tdew
 
-    Parameters
-    ----------
-    df:
-        Dataframe containing daily timeseries of
-        cloud cover fraction, tfmax, swe, and
-        shortwave radiation
-    sg:
-        Solar geometry dictionary, calculated with
-        `metsim.physics.solar_geom`.
-    pa:
-        Air pressure in Pascals
-    pva:
-        Vapor presure in Pascals
-    dtr:
-        Daily temperature range
-    params:
-        A dictionary of parameters from a MetSim object
+    this function must be vectorized and applied to each timestep
+    '''
 
-    Returns
-    -------
-    (tdew, svp):
-        A tuple of dewpoint temperature and saturation
-        vapor pressure
-    """
-    tt_max0 = sg['tt_max0']
-    potrad = sg['potrad']
-    daylength = sg['daylength']
-    yday = df.index.dayofyear - 1
-
-    t_tmax = np.maximum(tt_max0[yday] + (cnst.ABASE * pva), 0.0001)
-    t_final = t_tmax * df['tfmax']
-    df['pva'] = pva
-    df['tfinal'] = t_final
-
-    # Snowpack contribution
-    sc = np.zeros_like(df['tfmax'])
-    if params['mtclim_swe_corr']:
-        inds = np.logical_and(df['swe'] > 0.,  daylength[yday] > 0.)
-        sc[inds] = ((1.32 + 0.096 * df['swe'][inds]) *
-                    1.0e6 / daylength[yday][inds])
-        sc = np.maximum(sc, 100.)
-
-    # Calculation of shortwave is split into 2 components:
-    # 1. Radiation from incident light
-    # 2. Influence of snowpack - optionally set by MTCLIM_SWE_CORR
-    df['shortwave'] = potrad[yday] * t_final + sc
-
-    # Calculate cloud effect
+    tdew_old = t_min
+
+    while(np.sqrt(np.mean((tdew_temp-tdew_old)**2)) > params['tdew_tol']):
+        tdew_old = tdew_temp.copy()
+        vp = svp(tdew_temp)
+        sw = calc_shortwave(yday, tfmax, vp, sg)
+        pet = _calc_pet(sw, t_day, dayl, params)
+        tdew_temp = calc_tdew(pet, t_min, seasonal_prec, dtr)
```

Comment:
this function needs more work, previously the RMSE was calculated over the full timeseries so we'll need to work that in here.